Fix decomp behavior in export

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134657
* __->__ #132901

Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

tugsbayasgalan committed:
Fix decomp behavior in export

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"




[ghstack-poisoned]

angelayi commented on torch/export/exported_program.py:
```diff
@@ -345,6 +346,7 @@ def _decompose_and_get_gm_with_new_signature_constants(
 
         fake_args_unwrapped = pytree.tree_unflatten(fake_args, mod._in_spec)
         fake_mode = detect_fake_mode(fake_args)
+        fake_mode = contextlib.nullcontext() if fake_mode is None else fake_mode
```
Doesn't this make the code the same as the else case? where ep.verifier.dialect != "TRAINING"?

tugsbayasgalan replied:
The only difference is how we first find the fake_args. I plan to merge these two logic together somehow. Just testing CI first.

bdhirsh replied:
+1

tugsbayasgalan commented on torch/_subclasses/fake_impls.py:
```diff
@@ -373,6 +373,7 @@ def repeat_interleave_tensor(fake_mode, func, repeats, output_size=None):
     return repeats.new_empty(output_size)
 
 
+@register_op_impl(torch.ops.aten.item.default)
```
@bdhirsh What do you think about this? Because aten.item is a CIA fiunctional op that can be preserved, it needs to have direct meta implementation... This seems to be the easiest way to do it imo

bdhirsh replied:
Hmm I'm not sure why this is necessary. the `Meta` is part of the `CompositeImplicitAutograd` alias set, so `aten.item` should "automatically" have a registration to the meta key which is just "run the CIA impl"

bdhirsh replied:
confirmed with command below (you can see a registration to the Meta key):
```
>>> print(torch._C._dispatch_dump_table("aten::item"))
Undefined: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
CPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
CUDA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
HIP: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
XLA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
MPS: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
IPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
XPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
HPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
VE: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
Lazy: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
MTIA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
PrivateUse1: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
PrivateUse2: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
PrivateUse3: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
Meta: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
FPGA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
MAIA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
Vulkan: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
Metal: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedCPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedCUDA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedHIP: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedXLA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedMPS: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedIPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedXPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedHPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedVE: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedLazy: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedMTIA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedPrivateUse1: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedPrivateUse2: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedPrivateUse3: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
QuantizedMeta: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
CustomRNGKeyId: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
MkldnnCPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCUDA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseHIP: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseXLA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseMPS: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseIPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseXPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseHPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseVE: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseLazy: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseMTIA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparsePrivateUse1: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparsePrivateUse2: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparsePrivateUse3: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseMeta: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrCPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrCUDA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrHIP: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrXLA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrMPS: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrIPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrXPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrHPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrVE: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrLazy: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrMTIA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrPrivateUse1: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrPrivateUse2: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrPrivateUse3: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
SparseCsrMeta: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorCPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorCUDA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorHIP: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorXLA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorMPS: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorIPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorXPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorHPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorVE: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorLazy: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorMTIA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorPrivateUse1: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorPrivateUse2: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorPrivateUse3: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
NestedTensorMeta: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
BackendSelect: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/core/BackendSelectFallbackKernel.cpp:3 [backend fallback]
Python: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/core/PythonFallbackKernel.cpp:153 [backend fallback]
FuncTorchDynamicLayerBackMode: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/functorch/DynamicLayer.cpp:497 [backend fallback]
Functionalize: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
Named: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/core/NamedRegistrations.cpp:11 [kernel]
Conjugate: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/ConjugateFallback.cpp:17 [backend fallback]
Negative: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/native/NegateFallback.cpp:18 [backend fallback]
ZeroTensor: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/ZeroTensorFallback.cpp:86 [backend fallback]
ADInplaceOrView: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/core/VariableFallbackKernel.cpp:96 [backend fallback]
AutogradOther: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradCPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradCUDA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradHIP: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradXLA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradMPS: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradIPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradXPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradHPU: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradVE: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradLazy: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradMTIA: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradPrivateUse1: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradPrivateUse2: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradPrivateUse3: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradMeta: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
AutogradNestedTensor: registered at /home/hirsheybar/local/b/pytorch/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6903 [math kernel]
Tracer: registered at /home/hirsheybar/local/b/pytorch/torch/csrc/autograd/generated/TraceType_1.cpp:16082 [kernel]
AutocastCPU: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/autocast_mode.cpp:209 [backend fallback]
AutocastXPU: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/autocast_mode.cpp:351 [backend fallback]
AutocastCUDA: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/autocast_mode.cpp:165 [backend fallback]
FuncTorchBatched: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/functorch/BatchRulesDynamic.cpp:66 [kernel]
BatchedNestedTensor: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/functorch/LegacyBatchingRegistrations.cpp:758 [backend fallback]
FuncTorchVmapMode: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/functorch/VmapModeRegistrations.cpp:27 [backend fallback]
Batched: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/LegacyBatchingRegistrations.cpp:1075 [backend fallback]
VmapMode: fallthrough registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/VmapModeRegistrations.cpp:33 [backend fallback]
FuncTorchGradWrapper: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/functorch/TensorWrapper.cpp:207 [backend fallback]
PythonTLSSnapshot: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/core/PythonFallbackKernel.cpp:161 [backend fallback]
FuncTorchDynamicLayerFrontMode: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/functorch/DynamicLayer.cpp:493 [backend fallback]
PreDispatch: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/core/PythonFallbackKernel.cpp:165 [backend fallback]
PythonDispatcher: registered at /home/hirsheybar/local/b/pytorch/aten/src/ATen/core/PythonFallbackKernel.cpp:157 [backend fallback]
```

tugsbayasgalan replied:
Hmm so the issue here is we do need to run the python CIA impl here because this op produces unbacked symints. In our current logic, we don't register anything to CIA so as a result, it doesn't run the "analyzing unbacked symint".

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -411,18 +413,22 @@ def _decompose_and_get_gm_with_new_signature_constants(
-            constant_attrs,
-            _check_autograd_state=False,
-        )
+        with _ignore_backend_decomps(), fake_mode, _override_composite_implicit_decomp(
```
this is also now duped with the non-training path, I think?

tugsbayasgalan replied:
Hmm there is a bit of problem here. Even though the context manager is the same, we call different mechanism to get the aten IR. In the training IR case, we actually need sort of "retracing" kind of way (hence export_to_aten_ir), but in the non-training path, we call aot_aexport_module directly. I am still thinking about how to refactor this properly tho.

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

ydwu4 commented on torch/_export/utils.py:
```diff
@@ -42,6 +41,50 @@
 }
 
 
+def _get_shape_env(gm):
```
type annotation, maybe also change the name to '_get_shape_env_from_graph_module' or something?

tugsbayasgalan replied:
Soudns good!

ydwu4 commented on torch/_export/utils.py:
```diff
@@ -42,6 +41,50 @@
+    fake_mode = _detect_fake_mode_from_gm(gm)
+    if fake_mode is not None:
+        return fake_mode.shape_env
+    for v in vals:
```
Should we also assert that these shape_envs matches? Just to add more invariants to our system.

tugsbayasgalan replied:
Yeah good idea!

ydwu4 commented on torch/_export/utils.py:
```diff
@@ -42,6 +41,50 @@
+
+
+def _rename_without_collisions(
+    name_map: Dict[str, str],
```
Nit: Is this name_map closer to name_cache? Maybe we can rename it this way and reorder it to the end of call list like:
```
_rename_without_collisions(orig_name, new_name, is_placeholder, name_cache)
```
Seems more natural to me haha.

ydwu4 commented on torch/_export/utils.py:
```diff
@@ -42,6 +41,50 @@
+):
+    """
+    Renames nodes to avoid name collisions, with suffixing.
+    name_map: map from original name to new name
```
Probably also make this comment clearer like it's a cache for used names or something.

ydwu4 commented on torch/_export/utils.py:
```diff
@@ -475,9 +518,51 @@ def _bind_signature_to_inputs(mod, fake_args, fake_kwargs):
+                body_graph, array, args = node._args
+                subgraph_ph_tuples.append(
+                    (getattr(gm, body_graph.target), array + args)
+                )
```
probably error out when the hop not supported?

ydwu4 commented on torch/export/exported_program.py:
```diff
@@ -338,295 +269,139 @@ def _decompose_and_get_gm_with_new_signature_constants(
-                            fqn,
-                            mod_cls.__module__ + "." + mod_cls.__qualname__,
-                        )
+    # the exported module will store constants & non-persistent buffers such that
+    # retracing treats them as persistent buffers, so we inform the constants lifting pass
+    # and overwrite the new graph signature using the previous program.
+    constant_attrs = ConstantAttrMap()
+    non_persistent_buffers = {
+        spec.target
+        for spec in ep.graph_signature.input_specs
+        if spec.kind == InputKind.BUFFER and not spec.persistent
+    }
+    for name, value in ep.constants.items():
+        if name in non_persistent_buffers:
+            continue
+        # recursive getattr
+        _mod = mod
+        *atoms, attr = name.split(".")
+        for atom in atoms:
+            _mod = getattr(_mod, atom)
+        # remove as buffer, reassign as constant/non-persistent buffer
+        _mod._buffers.pop(attr, None)
+        setattr(_mod, attr, value)
+        constant_attrs.add(value, name)
```
Make it a helper function? So the data-dependency is clearer and abstracts away the details.

ydwu4 commented on torch/export/exported_program.py:
```diff
@@ -338,295 +269,139 @@ def _decompose_and_get_gm_with_new_signature_constants(
 
-    new_graph_signature = ExportGraphSignature(
-        input_specs=input_specs, output_specs=output_specs
-    )
-    # NOTE: aot_export adds symint metadata for placeholders with int
-    # values; since these become specialized, we replace such metadata with
-    # the original values.
-    # Also, set the param/buffer metadata back to the placeholders.
-    for old_node, new_node in zip(old_placeholders, new_placeholders):
-        if not isinstance(old_node.meta["val"], torch.Tensor):
-            new_node.meta["val"] = old_node.meta["val"]
-
-        if (
-            new_node.target in new_graph_signature.inputs_to_parameters
-            or new_node.target in new_graph_signature.inputs_to_buffers
-        ):
-            for k, v in old_node.meta.items():
-                new_node.meta[k] = v
+    for node in gm.graph.nodes:
+        if node.op == "placeholder":
+            if node.target in new_graph_signature.inputs_to_parameters:
+                param_name = new_graph_signature.inputs_to_parameters[node.target]
+                if param_name in params_buffers_to_node_meta:
+                    for k, v in params_buffers_to_node_meta[param_name].items():
+                        node.meta[k] = v
+            if node.target in new_graph_signature.inputs_to_buffers:
+                buffer_name = new_graph_signature.inputs_to_buffers[node.target]
+                if buffer_name in params_buffers_to_node_meta:
+                    for k, v in params_buffers_to_node_meta[buffer_name].items():
+                        node.meta[k] = v
+
+    # overwrite signature for non-persistent buffers
+    for spec in new_graph_signature.input_specs:
+        if spec.kind == InputKind.BUFFER and spec.target in non_persistent_buffers:
+            spec.persistent = False
```
Rewerite these for loops in to functions if possible? Ideally, the code could looks like the below _verify* calls.

ydwu4 reviewed (approved):
Looks good to me. Left some comments related to code organization.

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

bdhirsh commented on test/onnx/onnx_test_common.py:
```diff
@@ -380,6 +380,8 @@ def run_ort(
     else:
         ort_model = onnx_model
 
+    breakpoint()
```
🤔

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1370,26 +1350,9 @@ def _strict_export_lower_to_aten_ir(
-                if buffer_name in params_buffers_to_node_meta:
-                    for k, v in params_buffers_to_node_meta[buffer_name].items():
-                        node.meta[k] = v
+    _populate_param_buffer_metadata_to_new_gm(
```
Just created a utility function to share logic with run_decompositions

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1414,6 +1377,17 @@ def _strict_export_lower_to_aten_ir(
     _rename_constants_nodes(gm, export_graph_signature)
 
+    # Prettify names for placeholder nodes.
+    placeholder_naming_pass(
```
We do this because there are some param/buffer handling before this line, and this pass relies on the fact that all the param/buffers are settled before calling it.

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1742,6 +1716,18 @@ def _produce_guards_callback(gm):
     assert out_spec is not None
+
+    # Prettify names for placeholder nodes.
+    placeholder_naming_pass(
```
We do this because there are some param/buffer handling before this line, and this pass relies on the fact that all the param/buffers are settled before calling it.

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -247,227 +263,9 @@ def _(*args, **kwargs):
             decomp_table[op] = decomp
 
 
-def _rename_without_collisions(
```
Moved them to util file

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

pianpwk commented on torch/export/_reorder_placeholder_same_as_original_ep_pass.py:
```diff
@@ -0,0 +1,72 @@
+    not guaranteed to be same.
+    """
+    node_to_metadata = {}
+    original_order_to_node_name: Dict[int, str] = {}
```
nit: this can be a list, and remove cur_idx

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).

tugsbayasgalan committed:
Update on "Fix decomp behavior in export"


Previously, we called "_export_to_aten_ir" on training path because we needed to actually retrace the ep because training IR is not-functional. In the inference decompositions, we used to directly call aot_export_module with some wrapper code around. As we add more stuff to our code, it would mean we have to maintain two separate lowering code. Since inference decomp is subset of training decomp, in this PR, i just merge them together to avoid further code divergence. "_export_to_aten_ir" is bit of a overkill for inference decomp, but i think it is worth it to avoid confusions in the future. Also did some shuffling around on utility functions, so that code is little more organized. 

Differential Revision: [D60918293](https://our.internmc.facebook.com/intern/diff/D60918293)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60918293).
